Transactions Download functionality Improvements

jessy-p opened the pull request:
Fixes #10255 which is part of the Epic #9969 

#### Summary

The PR moves Transaction exports to a single implementation via Transact Platform Service API. This does away with the redundant JS based export for a single page export, and email + server export for larger data, that led to data inconsistency and additional maintenance overhead. With this PR, the merchant gets a browser download via an async service request, if it takes within 6 seconds ( 1 + 6 retries at 1 second interval ). There is always an email sent with the export file as a backup. If the file takes longer than 6 seconds to generate, the merchant will anyway get it on email.  

For security reasons, we are unable to store the export file in the uploads folder on the server. We were guided by the systems team to asynchronously poll the job id of the export and use that to get the URL of the exported report.

#### Changes proposed in this Pull Request
- removes the `@woocommerce/csv-export` based JS export on the Transactions page
- Uses the server `transactions/download` endpoint for all downloads instead.
- The export API returns an `export_id` 
- Polls the `transactions/download/<export_id>` endpoint to get the download URL.
- If the file is available within 6 retries ( 6 seconds) , the file is downloaded immediately and additionally emailed to the merchant. Else, it will be emailed to the merchant once the file gets ready.

By removing the browser export and using only server based export, the functionality is simplified and made consistent. We also preserve the functionality for the merchant by making the file available immediately in the browser whenever possible.

#### Testing instructions

- The necessary API changes to the Transact Platform Service is already merged
- Checkout this branch on Pressable test site via Jetpack Beta, or Local + Sandbox.
- Browse to `Payments > Transactions`
- Click the Export button in the top right corner of the report list
- You should see a snack bar notice with copy and appearance similar to this screenshot: 
![image](https://github.com/user-attachments/assets/2173187c-1727-4f8a-a145-af1207f8c966)

- Download various numbers of rows or filtered rows and check that the file is downloaded. An email is also sent to the merchant with the download URL.
- If the cron job is not running/slow, i.e the file is not available, the file will not be downloaded and a message displayed that the file will be emailed. When the job runs, the email is sent.

Here is a video of the complete flow ( please watch in full screen since the snack bar notice is obscured by the player controls in small screen ):

https://github.com/user-attachments/assets/d1fa9b31-fca1-4e0d-b533-5af24f6826ee


-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
Initial commit - a polling and fallback implementation on the export job

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10211` or branch name `update/9969-poll-async-job` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=update/9969-poll-async-job) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 054528354046f7e5223591f650013fd8f517560c
- Build time: 2025-02-07 16:07:42 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Jinksi commented on client/transactions/list/index.tsx:
```diff
@@ -670,16 +684,24 @@ export const TransactionsList = (
-					)
-				);
+				if ( response.export_id ) {
+					setTimeout( async () => {
```
We could use an interval to continue to poll the endpoint to see if the file is ready, e.g. check every 1s to a maximum of 5s.

jessy-p replied:
Confused about how much wait time is ideal.

Jinksi replied:
Me too. I'm thinking we check every ~1s, and set a max retries param.

I've explored this in #10212

jessy-p replied:
Merged back into this PR & resolving

Jinksi commented on client/transactions/list/index.tsx:
```diff
@@ -579,6 +583,16 @@ export const TransactionsList = (
+		createNotice( 'success', sprintf( noticeMessage, userEmail ) );
+	};
+
 	const endpointExport = async () => {
```
I think this is a great case for a React hook
* we're intending to share this across other reporting areas
* the logic is complex, with settimeout side-effects that might need cleaning up and error handling. I think this will be easier to reason about in a dedicated React hook.


I'll prepare an example of what it might look like to see what we think.

Jinksi replied:
I've opened a PR based on this branch in #10212 to explore the idea of using a reusable React hook to encapsulate the export request → polling → download logic.

Take a look and let me know what you think. Is it easier to reason about?

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
added changelog

Jinksi committed:
Add useReportExport hook to encapsulate polling logic for PR 10211 (#10212)

jessy-p commented on client/hooks/use-report-export.ts:
```diff
@@ -0,0 +1,177 @@
+				setIsDownloading( false );
+				createNotice(
+					'error',
+					__(
```
@Jinksi  It just means the job is slow / URL not available. At this point, it should say "Your export will be emailed ..". Fixed

Commit:
Bug fixes & Changes for DownloadURLResponse

- The response is changed to an array
- If the file times out, don't show an error. The file may be emailed.

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Deleted tests relating to the removed JS CSV download functionality.

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Added checks for status in response.

Commit:
Merge branch 'develop' into update/9969-poll-async-job

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Should handle the error if downloadURL API sends 500 error.

Commit:
Merge branch 'update/9969-poll-async-job' of https://github.com/Automattic/woocommerce-payments into update/9969-poll-async-job

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Removed some unused variables in the test file.

shendy-a8c committed:
Merge branch 'develop' into update/9969-poll-async-job

shendy-a8c commented on client/transactions/list/index.tsx:
```diff
@@ -579,13 +577,17 @@ export const TransactionsList = (
-		const { page, path, ...params } = getQuery();
-		const userEmail = wcpaySettings.currentUserEmail;
+	const onDownload = async () => {
+		recordEvent( 'wcpay_transactions_download_csv_click', {
```
Now, we don't pass `download_type` for event `wcpay_transactions_download_csv_click`.

I see by definition, [`download_type` is not nullable](https://github.com/Automattic/tracks-events-registration/blob/4f50833e576fdd98783c2f4de54f175e24164a20/events/wcadmin_wcpay_transactions_download_csv_click.event.json#L18). Would *not* passing it cause a problem? Should we change the definition to mark `download_type` as nullable?

haszari replied:
> Would not passing it cause a problem? Should we change the definition to mark download_type as nullable?

Shouldn't cause a problem, but I agree, we should make it optional (i.e. remove the prop, since it's not really relevant). 

Side note – I don't think `Download` is quite accurate for the button + event. This button (and event) is more like `Export` now.

jessy-p replied:
Made `download_type` nullable in [Event updates for Transactions CSV Export changes.](https://github.com/Automattic/tracks-events-registration/pull/2718#top)#2718

Commit:
Add test for report export hook

shendy-a8c reviewed (approved):
I have a hard time verifying that this change works, which is weird because couple days ago, when I reviewed https://github.com/Automattic/transact-platform-server/pull/7154, I also tested using this PR for the client side and I could see the file got downloaded after polling.

Some caveats:
- Since the [code checks if the download URL consists of `exports.wordpress.com`](https://github.com/Automattic/woocommerce-payments/blob/57378fc7299b0ee7238650fa6d5c4be3540fa793/client/hooks/use-report-export.ts#L85-L87), when testing locally, I had to comment that part out.
- I also increase the timeout interval to give me time to keep refreshing my WCPay server so the async job that generates the download file executes.

Check my screen recording, it shows:
- The above code changes locally.
- I clicked the download button.
- I refreshed my WCPay server page and shows that the email had been sent, indicating that the download was successfully.
- It shows the `Your CSV export is ready to download` notice.
- It shows under the dev console's Network tab that the CSV file is accessed.
- It shows that my download history is empty. The file was not downloaded.

I have a suspicion that maybe it's because my WCPay server's URL is http and Chrome's security just blocks it. Accessing the URL directly works, ie it downloads the file just fine.


https://github.com/user-attachments/assets/35211157-2637-4b8e-8619-2c40b009c624

Since I have a high confidence that this would work on production, I'm approving.

What is weird is that couple days ago, I could swear that the file got downloaded just fine.

nagpai commented:
I added a few tests for the newly added csv export hook  - [57378fc](https://github.com/Automattic/woocommerce-payments/pull/10211/commits/57378fc7299b0ee7238650fa6d5c4be3540fa793).  It would be great if @Jinksi can give this a look. A good bit of it was generated using cursor, and then fine-tuned by me, but I have a hunch it would need improvements :D.

Switching the status back to `needs review`

Commit:
Merge branch 'develop' into update/9969-poll-async-job

brucealdridge commented on client/transactions/list/index.tsx:
`labelInCsv` shouldn't be required anymore

nagpai replied:
The UI shows time in website time zone and export still shows UTC time zone. Can you elaborate why this would not be needed?

brucealdridge replied:
the `labelInCSV` prop here should be used for the column header in the client-side generated CSV. This is no longer necessary with the server-side export.

jessy-p replied:
Implemented [568023b](https://github.com/Automattic/woocommerce-payments/pull/10211/commits/568023b7dc002d5a8fa35ffd2d531ec35d402118) to remove the `labelInCsv` prop

brucealdridge reviewed (approved):
Have tested locally which didn't allow for download (even though a URL was returned) but emailing worked everytime.

Tested on a demo site and worked as expected. ✅ 

I thought this change was for all CSV's but only for transactions.

Design feedback:
the notice that pops up to say "Your download is being prepared" is easy to miss. likewise with the snackbar notice to say download is ready
![Screenshot 2025-01-30 at 2 53 58 PM](https://github.com/user-attachments/assets/25b39e6f-0214-41f1-80d7-375b93b256bd)
![Screenshot 2025-01-30 at 2 53 54 PM](https://github.com/user-attachments/assets/afbfa5d7-a7af-431d-9625-1ce51b8af8d9)

Jinksi commented:
> Design feedback:
the notice that pops up to say "Your download is being prepared" is easy to miss. likewise with the snackbar notice to say download is ready

IMHO I think this PR would benefit from a design review before we merge, to ensure we're using these elements correctly and giving users the appropriate UI feedback on their actions. I won't ping directly here, in case I've missed a prior discussion about this on P2 etc.

Commit:
Merge branch 'develop' into update/9969-poll-async-job

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Remove labelInCsv as it was used only for the
JS download functionality.

Commit:
Added a new event wcpay_transactions_download_csv_in_browser. Registered
via PR #2718 in tracks-events-registration repo.

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

jessy-p commented:
@rogermattic Could you please review the design. Will share the test site.

Main changes:

- While clicking on the Download button, we will show this notice:
<img width="876" alt="image" src="https://github.com/user-attachments/assets/be1dc9d1-dde4-4f6b-a4a4-484e8378fe15" />

- If the CSV is available for download within a short interval, we show this notice at the bottom of the page.
<img width="447" alt="image" src="https://github.com/user-attachments/assets/ae4114eb-a420-474e-905c-8b3150cdb5db" />

- If it is not available for download, we will show a similar notice that the file will be emailed.

haszari commented on client/hooks/use-report-export.ts:
```diff
@@ -0,0 +1,202 @@
+				'success' === exportedFileURLResponse.status &&
+				exportedFileURLResponse.download_url &&
+				exportedFileURLResponse.download_url.includes(
+					'exports.wordpress.com'
```
Why check `exports.wordpress.com` here? 

- Presumably this breaks local testing.
- If this is a security thing, can you add comment to explain why?

haszari replied:
> If this is a security thing

Actually that comment is useful either way!

jessy-p replied:
> Presumably this breaks local testing.

Does not break local testing, as exports.wordpress.com is hardcoded for every environment - https://github.com/Automattic/transact-platform-server/blob/1efcf99114e6539942d8e9738ea93a6e3638f9aa/server/wp-content/rest-api-plugins/endpoints/wcpay/service/class-export-service.php#L24

> security thing

I think it makes sure that we are downloading only a file from exports.wordpress.com, which would be placed there by the server. @Jinksi added this code I believe, do you have any suggestions?

Jinksi replied:
> @Jinksi added this code I believe, do you have any suggestions?

Git blame shows @jessy-p 😉 in https://github.com/Automattic/woocommerce-payments/commit/af1565e317e23fc8f4b82e4a1adffdeb967b3452.

I think it's fine to check only for `download_url` in the response, but agree that a comment would be helpful otherwise.

jessy-p replied:
Ah sorry, I don't remember - too long running PR, was going by the fact that it was the hook file from #10212. I'll add a comment.

jessy-p replied:
Oh but then [this](https://github.com/Automattic/woocommerce-payments/pull/10211#issuecomment-2632724006)

> Deleting the exports.wordpress.com seems to work, I can test locally now! I suspect there is a good reason for this check though.

I guess we can remove the check then.

haszari replied:
I'd like to better understand the motivation / need for the check, if it's important for real users (e.g. safety/security) then we should find a way to keep it AND allow flexible local testing 😁

jessy-p replied:
I've removed the check. I don't think its important for real users, the URL is returned from an API call that uses all our authentication mechanism like any other API from the WooCommerce Payments dashboard.

haszari commented on client/hooks/use-report-export.ts:
```diff
@@ -0,0 +1,202 @@
+					'?force_download=true';
+				link.click();
+
+				recordEvent( 'wcpay_transactions_download_csv_in_browser' );
```
Is this event useful to track? It's not really a user action. What questions could we answer with this event?

jessy-p replied:
We can see how many users get a downloaded file which is the aim of all this tediously implemented polling. Maybe all files end up only being emailed and we can detect this from this event. 

Can remove the event if it is mandatory that every event should only be a user action.

haszari replied:
I'd recommend removing it unless that question is important to answer.

jessy-p replied:
Removed the event as it is not user triggered and does not convey useful information about the user intention.

haszari commented on client/hooks/use-report-export.ts:
```diff
@@ -0,0 +1,202 @@
+					'success',
+					sprintf(
+						__(
+							'Your CSV export is ready to download. It will also be emailed to %s',
```
Need to review this wording & overall flow / experience. 

- `is ready` is wrong tense if it's already downloaded. 

Maybe `Your CSV export is ready and has been emailed to <email>! The file should download automatically; if not, click here to download directly.`.

rogermattic replied:
Could we (like I [suggested here](https://github.com/Automattic/woocommerce-payments/pull/10211#discussion_r1940929309))  just fire off this snackbar instantly when the download is initiated?


I'm trying to understand what exact case are we trying to support here:

> "Maybe Your CSV export is ready and has been emailed to <email>! The file should download automatically; if not, click here to download directly.." 

Would that be the user waiting to get an email, then going to check their email and then going back to the dashboard to see the notice and download it directly in the browser anyway?

I think a URL is a good option as a backup if only an in-browser download is offered.

haszari commented on client/hooks/use-report-export.ts:
```diff
@@ -0,0 +1,202 @@
+				link.href =
+					exportedFileURLResponse.download_url +
+					'?force_download=true';
+				link.click();
```
Will this work 100% reliably in all browsers? A fallback manual download link would be useful if it doesn't work for some reason.

jessy-p replied:
This is the existing logic used for downloading the file on this page  (via the library) as well as directly used in  other pages like Dispute Evidence - https://github.com/Automattic/woocommerce-payments/blob/4405c97033b6c3c9b1809bb99415a71d188d1aa7/client/payment-details/dispute-details/evidence-list.tsx#L72 

No issues have been raised so far regarding it having problems on any browser. Recently @Jinksi improved the code with this [change](https://github.com/Automattic/transact-platform-server/pull/7096)

haszari replied:
I suspect it works well 99.99% of the time, I'm just wondering if a clickable link would be useful/preferable for some users. No biggie, happy to proceed with fully automatic download 👌🏻

haszari commented on client/hooks/use-report-export.ts:
```diff
@@ -0,0 +1,202 @@
+					'success',
+					sprintf(
+						__(
+							'Your export will be emailed to %s',
```
Need to review this wording too – more clearly explain what's happening. 

Something like `Your CSV export is being processed. It will be sent to <email> within 24 hours`.

jessy-p replied:
This is the earlier functionality, unchanged except for moving it to a different file. Can we take this up in a new issue ?

rogermattic replied:
I don't think we need to scare the users with the 24h waiting period (I assume realistically in major cases it will happen faster). I [suggested some improvements here](https://github.com/Automattic/woocommerce-payments/pull/10211#discussion_r1940929309).

haszari replied:
> I don't think we need to scare the users with the 24h waiting period (I assume realistically in major cases it will happen faster).

For sure, that is a placeholder for whatever a good value is. Sounds like we're broadly aligned @rogermattic it's just getting the details of the flow and copy right 👌🏻

haszari commented on client/hooks/use-report-export.ts:
```diff
@@ -0,0 +1,202 @@
+ * } );
+ */
+export const useReportExport = () => {
+	const [ isDownloading, setIsDownloading ] = useState( false );
```
Does this state reflect "downloading" or that a CSV request is active/processing? 

Might be clearer to rename. Is "in progress" accurate (i.e. keep it simple!)?

Jinksi replied:
As the person who wrote this variable name, I totally agree that `isDownloading` is not explicit enough and misleading.

* The actual file download happens instantaneously when the browser triggers the download via the created link element
* Most of the time is spent waiting for the server to generate/prepare the export

Since this is a bit more complex that a simple request, maybe something more descriptive like `isExportInProgress` to better communicate _what_ is in progress – the export operation on server.

jessy-p replied:
Changed to `isExportInProgress`

haszari commented on client/transactions/list/index.tsx:
```diff
@@ -815,6 +763,14 @@ export const TransactionsList = (
+						'Your download is being prepared.',
+						'woocommerce-payments'
+					) }
+				</Notice>
```
Can we use a spinner / indeterminate progress instead of a `Notice` here?

- `Notice` is disruptive – moves the UI around.
- Tricky to get the wording right, the notice is effectively showing that we're processing.
- `warning` is not accurate 😁 nothing to worry about

Jinksi replied:
Yes we talked about making the download button an in-progress/spinner state instead but IIRC we opted to propose the current UI first and ask design for feedback.

rogermattic replied:
Just reviewing the test site @jessy-p put together!

> Can we use a spinner / indeterminate progress instead of a Notice here?

Not a big fan of a spinner as it blocks the user from doing anything else, which we don't need to do.  Though I agree with these points below:

> `Notice` is disruptive – moves the UI around.
> Tricky to get the wording right, the notice is effectively showing that we're processing.
> `warning` is not accurate 😁 nothing to worry about

I agree that a notice isn't the right component to use and comes with additional unnecessary challenges. From the components we have (even though this is not **_ideal_** but at least brings the consistency)  that we could use I'd suggest the [snackbars](https://wordpress.github.io/gutenberg/?path=/docs/components-snackbarlist--docs). 
<img width="520" alt="Screenshot 2025-02-04 at 11 36 40" src="https://github.com/user-attachments/assets/3557ce4b-b0c0-4655-a85e-0e84655dc9ef" />

<img width="1600" alt="Exporting a report – More than 50 rows selected – 3 and more email recipients" src="https://github.com/user-attachments/assets/bd617e90-d7a1-4dd5-80cf-60c4645c73c2" />
First showing the:

> _Your report download has started._

and just a short moment later:

> _Your report will be sent to [email] when it’s ready._ 

Or we could just say:

> _Your report will be sent to [email]._ 

To completely skip the confusing parts. 




These snackbars should be dismissable and stacked and then disappear after some time.

rogermattic replied:
Just trying to make sure I’ve covered everything with my comment above. Do I understand this correctly?

We have two cases:
	1.	The file is small, so it’s always downloaded in the browser and emailed to the user.
	2.	The file is big, so it’s only emailed?

Or do we always try to download in the browser and send an email regardless?

If we have these two cases, I'd say we should:

> 	1.	The file is small, so it’s always downloaded in the browser and emailed to the user.

Use these two snackbars:
>Your report download has started.

>Your report will be sent to [email].

and
>	2.	The file is big, so it’s only emailed

Use only a single snackbar:

> Your report will be sent to [email] when it’s ready.

If we unify the flows so that we always try to download in the browser and send an email regardless, but we’re concerned that it might take longer (both in the browser and for the email to arrive), then I think we should still use these two snackbars and display them sequentially:

> Your report download has started. This may take a moment.
> Your report will be sent to [email] when it’s ready.

This messaging acknowledges that it might take some time to receive the file. I don’t think we need to be more precise—users wouldn’t expect live updates anyway.

jessy-p replied:
Based on the ideas above, and the fact that we cannot/ don't make upfront decision anymore that the file is big or small, this would ALWAYS be

>  Your report download has started.

followed by

> Your report will be sent to [email].

We don't need to message the user that file is being downloaded, as that will be visible. We can also skip the URL as it is being emailed anyway. 

@rogermattic Hope above is correct?

@haszari

rogermattic replied:
To confirm, that's 100% how I see it. 

> We don't need to message the user that file is being downloaded, as that will be visible. We can also skip the URL as it is being emailed anyway.

I mentioned it [here](https://github.com/Automattic/woocommerce-payments/pull/10211#discussion_r1940947631) too.

haszari replied:
> I agree that a notice isn't the right component to use and comes with additional unnecessary challenges. From the components we have (even though this is not ideal but at least brings the consistency) that we could use I'd suggest the [snackbars](https://wordpress.github.io/gutenberg/?path=/docs/components-snackbarlist--docs).

@rogermattic These two snackbars look like a great experience, addresses my concerns 🙌 

Though … the order seems reversed to me!

> Based on the ideas above, and the fact that we cannot/ don't make upfront decision anymore that the file is big or small, this would ALWAYS be
> `Your report download has started.` > `Your report will be sent to [email].`

The process that's happening is more like this:

1. User clicks `download` (aka export) button.
2. Service starts generating a CSV. 
  - This takes seconds / moments – not instantaneous.
  - UI polls the service for a download URL.
  - aka "processing" time
3. Download URL is available, file starts downloading. 
4. Some time later, file shows up in email. 

Because the download doesn't start immediately, the `Your report download has started` snackbar doesn't feel right to me. That mismatch is problematic. 

I'm keen to make the snackbar messages / UI accurately reflect the flow. Something like:

1. User clicks `download` (aka export) button.
  - Snackbar `Your report will be emailed to admin@webstore.com when it's ready` - i.e. 2. Service starts generating a CSV. 
processing / acknowledge request
3. Download URL is available, file starts downloading. 
  - Maybe this doesn't need a snackbar? 
5. Some time later, file shows up in email. 

Maybe a heavy snackbar is a good fit, with everything:

`Processing your export. The file will download automatically and emailed to admin@webstore.com.`

Side question:

- Should we refer to these CSV files as `report` or `export` or something else? I don't like the term report, as there's no analysis, it's just a raw export of the data. I think of reports as summary stats, graphs etc. 
- Should we change the `Download` button to `Export`? 

@rogermattic @jessy-p what do you think?

haszari replied:
One more point to clarify – I think we _could_ use a spinner state on the `Download` button (@rogermattic  when I refer to spinner I mean any kind of indeterminate progress indicator, and in general, don't want to interrupt the user with modals!). 

E.g. 

1. User clicks `download` (aka export) button.
    - Service starts generating a CSV. 
    - `Download` [button spins, barber pole style - `isBusy`](https://developer.wordpress.org/block-editor/reference-guides/components/button/#isbusy). 
    - User can't click it while the export is processing – i.e. accidentally trigger multiple exports.
3. Download URL is available, file starts downloading. 
    - Snackbar: `Your report is downloading and will be sent toadmin@webstore.com`
    - `Download` button goes back to normal state
4. Some moments later, file shows up in email.

jessy-p replied:
Going with 

> Processing your export. The file will download automatically and emailed to admin@webstore.com.

Not showing any notice/snackbar at file download.

rogermattic replied:
Agree with @haszari’s suggestion to change ‘Download’ to ‘Export’ (I also made that swap in that recent design proposal).

nagpai replied:
I have renamed `Download` label to `Export` and also changed the icon to what appears on the recent design

![image](https://github.com/user-attachments/assets/6922f70e-fe16-4b9b-acca-057afcd774cb)


Commits: [4b815ae](https://github.com/Automattic/woocommerce-payments/pull/10211/commits/4b815aed1730b02473a5e91724fd3b873f0df258) , [3b84602](https://github.com/Automattic/woocommerce-payments/pull/10211/commits/3b846028b30b58b05efb295be38ae01eb83cddcb)

haszari reviewed (commented):
I think this is close but we need to do some more work on the UI / flow. Tricky to get this right without pulling a thread that leads to implementing more improvements to the flow!

- As noted by @brucealdridge and @shendy-a8c. The inline flow is not easy to test locally (I had a quick test and it didn't work). That's not ideal, we need to be able to easily test the same behaviour that merchants experience. If we can't, then that's a major compromise so need to be careful.
- The UI needs some polish & wording tweaks.
  - Use a spinner to show that the export is processing (not a `Notice`). 
  - More friendly and clear wording for the snackbar.
  - Download link fallback in case the auto download doesn't work. (Not sure if that's essential, it's a common pattern and puts user in control.)

With that in mind, I think we need to revisit if it's worth suporting polling / inline download now, or if we keep it simple and just always email.

I definitely support implementing a polling UI. Looking at the existing UI, and the subtleties of getting the flow right, I'm realising that we're treading a fine line trying to preserve the current experience, and not devote more time to building a better flow (e.g. based on @rogermattic designs). 

@jessy-p @nagpai what do you think? 

Also please keep the PR description updated with full details of the flow (screenshots of all states or a video, and all relevant copy as text). This makes it clear for reviewers what's changed/new, and makes it easy to review / feedback (e.g. for design & other busy stakeholders). 

Would be good to set up a demo site with this so we can review the full flow in context, since it's different in local env.

haszari commented:
> - Allow testing inline-download flow locally. 
> - Use a spinner to show that the export is processing.
> - More friendly and clear wording for the snackbar.
> - Download link fallback in case the auto download doesn't work. (Not sure if that's essential, it's a common pattern and puts user in control.)

I started a draft PR to explore tweaks to address feedback:

- #10292

> Allow testing inline-download flow locally. 

Deleting the `exports.wordpress.com` seems to work, I can test locally now! I suspect there is a good reason for this check though.

jessy-p commented:
> The inline flow is not easy to test locally

Can you switch to https locally ? I'm able to test locally.

shendy-a8c committed:
Merge branch 'develop' into update/9969-poll-async-job

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Address review comments

- Removed the file URL check that blocks local testing
- Changed isDownloading to isExportInProgress
- Removed the track event as it is not user triggered.

Commit:
Merge branch 'update/9969-poll-async-job' of https://github.com/Automattic/woocommerce-payments into update/9969-poll-async-job

Commit:
Changed to a single notice.

shendy-a8c committed:
Merge branch 'develop' into update/9969-poll-async-job

shendy-a8c committed:
Fix unit test after createNotice() call for success case is removed from useReportExport().

shendy-a8c commented:
[7e10a7d](https://github.com/Automattic/woocommerce-payments/pull/10211/commits/7e10a7d1a7983446b6f25b0fa8e6c2fdd00802df) fixes unit test due to changes from [d7e1f6d](https://github.com/Automattic/woocommerce-payments/pull/10211/commits/d7e1f6db6523df4e91c7867d839faad81815d8e6).

Jinksi commented:
✅ Failing e2e test is unrelated, due to the removal of the date settings notice in #10129:

<img width="939" alt="image" src="https://github.com/user-attachments/assets/2ea95077-8b5d-46ec-875e-a1b837889936" />

[playwright-report.zip](https://github.com/user-attachments/files/18682629/playwright-report.zip) ([from this test run](https://github.com/Automattic/woocommerce-payments/actions/runs/13167015716?pr=10211))

Jinksi committed:
Remove unused sprintf import from use-report-export hook

Jinksi committed:
Merge branch 'develop' into update/9969-poll-async-job

Jinksi committed:
Add documentation for export response interfaces in use-report-export hook

Jinksi committed:
Add isBusy prop to DownloadButton component for export progress indication

Jinksi commented on changelog/update-9969-poll-async-job:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: update
+
+Replaced JS export with downloading Transactions CSV from service.
```
Maybe this could be more descriptive, to explain the changes from a merchant's perspective.


```suggestion
Enhanced transaction CSV exports with async transact platform service processing, improving the CSV export experience.
```

(My attempt, please modify or ignore 😉)

nagpai replied:
Updated the change log with hints from what you have shared, Eric - [ef7b796](https://github.com/Automattic/woocommerce-payments/pull/10211/commits/ef7b796dc2b58a90248d96198a09becf3bc3f6e3)

Jinksi reviewed (approved):
This tests well! I've tested it on a Pressable site against with various filters enabled, all is working smoothly. It feels magic to have a transact-platform-server generated CSV appear so instantly as a client-side download 🌟 

Flicking my network off immediately after clicking download, to force the polling to fail, results in the CSV still being emailed, and the UI gracefully cancelling the polling process ✅

I've added some minor commits that:
* Remove unused `sprintf` import in cfead5691ee4f9cea74ce17dc89c1baec6c27ecc
* Add extra documentation for export response interfaces in b06884029233cd50b31ee0bc7e3dca7b6875cc84: I was getting confused reading `ExportResponse` and `ExportURLResponse` and thought that we can make it clearer what these response objects contain and their purpose in the export flow.
* Add `isBusy` prop to `DownloadButton` to show loading state during export in b52e61fa506677273681587b576d8ce45fda2996 - this provides better user feedback while the export is being processed, following WP component patterns for loading states. **If this change is incorrect, e.g. I missed a discussion about this already, please feel free to revert it** 👍


https://github.com/user-attachments/assets/ff0c9678-608e-4a80-be65-309985482cb5


Pulling some thoughts from our discussion in Slack (p1738811085595819-slack-C02BW3Z8SHK):

>Do you think the notice stays for too much time ?

No, I think it is a sufficient amount of time for the merchant to read the message, and the Snackbar can be dismissed on click.

> Is the polling working reliably (just once, I felt like it polled once, and failed - worked all the other times). I haven't changed anything about it, so not sure if it was just a issue with my build that one time. 

It has been reliable for me when testing on a Pressable site 👍

haszari commented:
> Add [isBusy](https://github.com/Automattic/woocommerce-payments/pull/10211#discussion_r1942014946) prop to DownloadButton to show loading state during export in https://github.com/Automattic/woocommerce-payments/commit/b52e61fa506677273681587b576d8ce45fda2996 - this provides better user feedback while the export is being processed, following WP component patterns for loading states.

🙇🏻  Thank you @Jinksi !

haszari commented on client/transactions/list/index.tsx:
```diff
@@ -646,88 +672,23 @@ export const TransactionsList = (
+				sprintf(
 					__(
-						'There was a problem generating your export.',
+						'Processing your export. The file will download automatically and will be emailed to %s',
```
This wording feels a little sad! Not a blocker but here are some ideas:

- Can we add a symbol / emoji to make it more happy / triumphant? 
- Add a full stop at end of sentence so it doesn't dangle 
- If possible, might be helpful to use different formatting for the email (bold or something).

> ✅ Now processing your export. The file will automatically download and will be emailed to **admin@store.com**.

nagpai replied:
I have added an icon and full stop. I was unable to add a bold font face for the email, since createNotice supports only string. 

<img width="654" alt="image" src="https://github.com/user-attachments/assets/08715bca-f09a-4395-91df-f2d27f462adb" />

rogermattic replied:
Hmmm @haszari @nagpai @jessy-p I feel like this first part of the message sounds a bit weird.. could we say: "We’re processing your export. The file will download automatically and be emailed to [email]."

In terms of the celebratory symbol and layout, could we remove the side padding and keep the style consistent with the [original component](https://wordpress.github.io/gutenberg/?path=/docs/components-snackbar--docs) without diverging too much? If we want, we could add 🚀 or 🎉 inline instead. 

For example:

> We’re processing your export. 🎉 The file will download automatically and be emailed to [email]. 
or 
> We’re processing your export. The file will download automatically and be emailed to [email]. 🚀  

I'm just thinking ✅ means _finished_. But if you want to keep it, then please, could it be inline?

haszari replied:
> We’re processing your export. 🎉 The file will download automatically and be emailed to [email].

Sounds good to me, thanks for following up and refining this @rogermattic ! 

Also works without the emoji, if that's preferable:

> We’re processing your export. The file will download automatically and be emailed to [email].

rogermattic replied:
Don't mind either way. If the current version in the code includes emoji, then leave it in :)

nagpai replied:
I am all for emoji :D !

nagpai replied:
I have created a quick issue to make sure we do not miss this improvement - #10356

haszari reviewed (approved):
Quick review - tested locally only. Looking good, we can maybe polish the flow a little more but we can do that as follow up.

- In my first test, the download didn't happen, and felt strange when the button was still spinning but the snackbar had disappeared. Would be nice if we can sync them, but this is maybe an edge case.
- The snackbar is a little sad, it should be be gently triumphant! 

<img width="763" alt="Screenshot 2025-02-07 at 3 57 35 PM" src="https://github.com/user-attachments/assets/1e976830-0ebe-4bc5-a115-aa1d3acd87ea" />

Commit:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Rename 'Download' button label to 'Export'

Commit:
Change icon on Export button

Commit:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Update snackbar notice

Commit:
Update changelog

Commit:
Merge branch 'develop' into update/9969-poll-async-job

nagpai commented:
Since we have approvals, and no more blocking feedback remains, I am merging this branch. It will also help speed up Disputes and Payout exports. If any major bugs are found, we can reopen.
